File: packages/sdk/src/sss3.ts
/**
 * SSS-3 SDK Module — Private Stablecoin Helpers
 *
 * Provides typed client methods for the SSS-3 preset:
 *   - Allowlist-gated transfers
 *   - Token-2022 ConfidentialTransfer extension scaffold
 *   - Auditor ElGamal pubkey management
 *
 * Usage:
 *   import { SSS3Client } from '@sss/sdk/sss3';
 *   const client = new SSS3Client(provider, idl);
 */

import {
  AnchorProvider,
  Program,
  BN,
  Idl,
} from "@coral-xyz/anchor";
import {
  Connection,
  Keypair,
  PublicKey,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
  Transaction,
  TransactionInstruction,
  TransactionSignature,
} from "@solana/web3.js";
import {
  getAssociatedTokenAddressSync,
  TOKEN_2022_PROGRAM_ID,
  ASSOCIATED_TOKEN_PROGRAM_ID,
  ExtensionType,
  getMintLen,
  createInitializeConfidentialTransferMintInstruction,
} from "@solana/spl-token";
import crypto from "crypto";

import {
  Sss3Params,
  Sss3Config,
  Sss3AllowlistRecord,
  AllowlistEntryParams,
  Sss3Info,
  AllowlistStatus,
} from "./types";
import { PDAs } from "./pdas";
import { encodeName, encodeSymbol, decodeString } from "./utils";
import { PROGRAM_ID } from "./client";

// ─── Types ────────────────────────────────────────────────────────────────────

export type ElGamalPubkey = Uint8Array; // 32 bytes

export interface ConfidentialMintProof {
  commitmentHash: Uint8Array; // SHA-256(recipient || amount || nonce)
  nonce: Uint8Array;          // 32-byte random nonce
}

export interface ConfidentialTransferProof {
  ciphertext: Uint8Array;     // ElGamal ciphertext bundle
  proof: Uint8Array;          // ZK range proof
}

// ─── Client ───────────────────────────────────────────────────────────────────

export class SSS3Client {
  readonly program: Program;
  readonly provider: AnchorProvider;
  readonly pdas: PDAs;

  constructor(provider: AnchorProvider, idl: Idl, programId?: PublicKey) {
    this.provider = provider;
    this.program = new Program(idl, programId ?? PROGRAM_ID, provider);
    this.pdas = new PDAs(programId ?? PROGRAM_ID);
  }

  get connection(): Connection {
    return this.provider.connection;
  }

  get wallet(): PublicKey {
    return this.provider.wallet.publicKey;
  }

  // ─── Initialization ──────────────────────────────────────────────────────────

  /**
   * Initialize a new Token-2022 mint with the ConfidentialTransfer extension.
   *
   * MUST be called before `initializeSss3()`. The mint account is pre-created
   * with the CT extension so that Anchor can validate the mint in the same
   * transaction as the program instruction.
   *
   * @param auditorElgamalPubkey - Optional 32-byte auditor ElGamal pubkey.
   *   When set, all confidential transfers include a secondary ciphertext
   *   encrypted to the auditor for regulatory visibility.
   * @returns The mint keypair and transaction signature
   */
  async initializeConfidentialMint(
    auditorElgamalPubkey?: ElGamalPubkey
  ): Promise<{ mint: Keypair; txSig: string }> {
    const mint = Keypair.generate();
    const extensions = [ExtensionType.ConfidentialTransferMint];
    const mintLen = getMintLen(extensions);
    const lamports =
      await this.connection.getMinimumBalanceForRentExemption(mintLen);

    const createAccountIx = SystemProgram.createAccount({
      fromPubkey: this.wallet,
      newAccountPubkey: mint.publicKey,
      space: mintLen,
      lamports,
      programId: TOKEN_2022_PROGRAM_ID,
    });

    const initCtMintIx = createInitializeConfidentialTransferMintInstruction(
      mint.publicKey,
      null,           // authority — null means immutable after init
      true,           // auto_approve_new_accounts
      auditorElgamalPubkey ?? null,
      TOKEN_2022_PROGRAM_ID
    );

    const tx = new Transaction().add(createAccountIx, initCtMintIx);
    const txSig = await this.provider.sendAndConfirm(tx, [mint]);

    return { mint, txSig };
  }

  /**
   * Initialize a new SSS-3 stablecoin.
   *
   * Call `initializeConfidentialMint()` first if `params.confidentialTransfersEnabled = true`
   * to pre-initialize the Token-2022 CT extension on the mint account.
   *
   * @param mintKeypair - Pre-generated mint keypair (or from initializeConfidentialMint)
   * @param params - SSS-3 initialization parameters
   */
  async initializeSss3(
    mintKeypair: Keypair,
    params: Sss3Params
  ): Promise<TransactionSignature> {
    const [configPDA] = this.pdas.config(mintKeypair.publicKey);
    const [sss3ConfigPDA] = this.pdas.sss3Config(mintKeypair.publicKey);

    const auditorPubkeyArg = params.auditorPubkey
      ? { some: Array.from(params.auditorPubkey) }
      : { none: {} };

    return await this.program.methods
      .initializeSss3({
        name: encodeName(params.name),
        symbol: encodeSymbol(params.symbol),
        decimals: params.decimals,
        supplyCap: params.supplyCap,
        allowlistAuthority: params.allowlistAuthority,
        requireAllowlistForReceive: params.requireAllowlistForReceive,
        requireAllowlistForSend: params.requireAllowlistForSend,
        confidentialTransfersEnabled: params.confidentialTransfersEnabled,
        autoApproveNewAccounts: params.autoApproveNewAccounts,
        auditorPubkey: auditorPubkeyArg,
      })
      .accounts({
        payer: this.wallet,
        mint: mintKeypair.publicKey,
        config: configPDA,
        sss3Config: sss3ConfigPDA,
        systemProgram: SystemProgram.programId,
        tokenProgram: TOKEN_2022_PROGRAM_ID,
        rent: SYSVAR_RENT_PUBKEY,
      })
      .signers([mintKeypair])
      .rpc();
  }

  // ─── Allowlist Management ─────────────────────────────────────────────────────

  /**
   * Add a wallet to the SSS-3 allowlist.
   * Only callable by the `allowlist_authority` set during initialization.
   */
  async allowlistAdd(
    mint: PublicKey,
    wallet: PublicKey,
    params: AllowlistEntryParams
  ): Promise<TransactionSignature> {
    const [configPDA] = this.pdas.config(mint);
    const [sss3ConfigPDA] = this.pdas.sss3Config(mint);
    const [allowlistPDA] = this.pdas.sss3Allowlist(mint, wallet);

    // Encode note to 64-byte fixed buffer
    const noteBytes = Buffer.alloc(64);
    Buffer.from(params.note, "utf8").copy(noteBytes, 0, 0, Math.min(64, Buffer.byteLength(params.note, "utf8")));

    return await this.program.methods
      .allowlistAdd(params.expiry, Array.from(noteBytes))
      .accounts({
        allowlistAuthority: this.wallet,
        config: configPDA,
        sss3Config: sss3ConfigPDA,
        wallet,
        allowlistRecord: allowlistPDA,
        systemProgram: SystemProgram.programId,
      })
      .rpc();
  }

  /**
   * Remove a wallet from the SSS-3 allowlist.
   * Sets `active = false` without closing the account (preserves audit trail).
   */
  async allowlistRemove(
    mint: PublicKey,
    wallet: PublicKey
  ): Promise<TransactionSignature> {
    const [configPDA] = this.pdas.config(mint);
    const [sss3ConfigPDA] = this.pdas.sss3Config(mint);
    const [allowlistPDA] = this.pdas.sss3Allowlist(mint, wallet);

    return await this.program.methods
      .allowlistRemove()
      .accounts({
        allowlistAuthority: this.wallet,
        config: configPDA,
        sss3Config: sss3ConfigPDA,
        wallet,
        allowlistRecord: allowlistPDA,
      })
      .rpc();
  }

  // ─── Transfer ────────────────────────────────────────────────────────────────

  /**
   * Transfer tokens with SSS-3 allowlist enforcement.
   *
   * Automatically includes sender/recipient allowlist record accounts based on
   * the `require_allowlist_for_send` / `require_allowlist_for_receive` flags.
   */
  async transferSss3(
    mint: PublicKey,
    from: PublicKey,
    to: PublicKey,
    amount: BN
  ): Promise<TransactionSignature> {
    const [configPDA] = this.pdas.config(mint);
    const [sss3ConfigPDA] = this.pdas.sss3Config(mint);
    const fromTokenAccount = getAssociatedTokenAddressSync(
      mint, from, false, TOKEN_2022_PROGRAM_ID
    );
    const toTokenAccount = getAssociatedTokenAddressSync(
      mint, to, false, TOKEN_2022_PROGRAM_ID
    );
    const [senderAllowlist] = this.pdas.sss3Allowlist(mint, from);
    const [recipientAllowlist] = this.pdas.sss3Allowlist(mint, to);

    return await this.program.methods
      .transferSss3(amount)
      .accounts({
        signer: this.wallet,
        config: configPDA,
        sss3Config: sss3ConfigPDA,
        mint,
        senderTokenAccount: fromTokenAccount,
        recipientTokenAccount: toTokenAccount,
        senderAllowlistRecord: senderAllowlist,
        recipientAllowlistRecord: recipientAllowlist,
        tokenProgram: TOKEN_2022_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .rpc();
  }

  // ─── Confidential Mint ────────────────────────────────────────────────────────

  /**
   * Generate a commitment hash for a confidential mint operation.
   *
   * commitment_hash = SHA-256(recipient_pubkey || amount_le_8 || nonce_32)
   *
   * This hash is recorded on-chain by `confidentialMintSss3()` to create an
   * auditable link between the authorization event and the actual token minting.
   */
  generateCommitmentHash(
    recipient: PublicKey,
    amount: BN,
    nonce?: Uint8Array
  ): ConfidentialMintProof {
    const resolvedNonce = nonce ?? crypto.randomBytes(32);
    const amountBuf = amount.toArrayLike(Buffer, "le", 8);

    const hash = crypto.createHash("sha256")
      .update(recipient.toBuffer())
      .update(amountBuf)
      .update(resolvedNonce)
      .digest();

    return {
      commitmentHash: new Uint8Array(hash),
      nonce: new Uint8Array(resolvedNonce),
    };
  }

  /**
   * Initiate a confidential mint:
   *  1. Validates the mint authority signature
   *  2. Validates recipient is on allowlist (active, non-expired)
   *  3. Records commitment_hash on-chain for auditor decryption
   *
   * After this instruction succeeds, call the Token-2022 `confidentialTransferMint`
   * instruction with the ElGamal ciphertext to complete the actual minting.
   */
  async confidentialMintSss3(
    mint: PublicKey,
    recipient: PublicKey,
    commitmentHash: Uint8Array
  ): Promise<TransactionSignature> {
    const [configPDA] = this.pdas.config(mint);
    const [sss3ConfigPDA] = this.pdas.sss3Config(mint);
    const [recipientAllowlist] = this.pdas.sss3Allowlist(mint, recipient);
    const recipientTokenAccount = getAssociatedTokenAddressSync(
      mint, recipient, false, TOKEN_2022_PROGRAM_ID
    );

    return await this.program.methods
      .confidentialMintSss3(Array.from(commitmentHash))
      .accounts({
        mintAuthority: this.wallet,
        config: configPDA,
        sss3Config: sss3ConfigPDA,
        mint,
        recipient,
        recipientAllowlistRecord: recipientAllowlist,
        recipientTokenAccount,
        tokenProgram: TOKEN_2022_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .rpc();
  }

  // ─── Read Methods ─────────────────────────────────────────────────────────────

  /**
   * Fetch the SSS-3 configuration for a mint.
   */
  async getSss3Config(mint: PublicKey): Promise<Sss3Config> {
    const [sss3ConfigPDA] = this.pdas.sss3Config(mint);
    const raw = await this.program.account.sss3Config.fetch(sss3ConfigPDA);

    return {
      mint: raw.mint as PublicKey,
      allowlistAuthority: raw.allowlistAuthority as PublicKey,
      requireAllowlistForReceive: raw.requireAllowlistForReceive as boolean,
      requireAllowlistForSend: raw.requireAllowlistForSend as boolean,
      confidentialTransfersEnabled: raw.confidentialTransfersEnabled as boolean,
      autoApproveNewAccounts: raw.autoApproveNewAccounts as boolean,
      auditorPubkey: raw.auditorPubkey
        ? new Uint8Array(raw.auditorPubkey as number[])
        : null,
      allowlistCount: raw.allowlistCount as number,
      bump: raw.bump as number,
    };
  }

  /**
   * Fetch full SSS-3 info (base config + SSS-3 config).
   */
  async getSss3Info(mint: PublicKey): Promise<Sss3Info> {
    const [sss3ConfigPDA] = this.pdas.sss3Config(mint);
    const config = await this.getSss3Config(mint);

    return {
      address: sss3ConfigPDA,
      mint,
      config,
      allowlistCount: config.allowlistCount,
    };
  }

  /**
   * Fetch an allowlist record for a wallet.
   * Returns null if the wallet has never been allowlisted.
   */
  async getAllowlistRecord(
    mint: PublicKey,
    wallet: PublicKey
  ): Promise<Sss3AllowlistRecord | null> {
    const [allowlistPDA] = this.pdas.sss3Allowlist(mint, wallet);
    try {
      const raw = await this.program.account.allowlistRecord.fetch(allowlistPDA);
      const noteBytes = Buffer.from(raw.note as number[]);
      // Trim null bytes from fixed-size note field
      const noteStr = noteBytes.toString("utf8").replace(/\0+$/, "");

      return {
        mint: raw.mint as PublicKey,
        wallet: raw.wallet as PublicKey,
        addedAt: raw.addedAt as BN,
        expiresAt: raw.expiresAt as BN,
        addedBy: raw.addedBy as PublicKey,
        active: raw.active as boolean,
        note: noteStr,
        bump: raw.bump as number,
      };
    } catch {
      return null;
    }
  }

  /**
   * Check the allowlist status for a wallet, including expiry validation.
   */
  async getAllowlistStatus(
    mint: PublicKey,
    wallet: PublicKey
  ): Promise<AllowlistStatus> {
    const record = await this.getAllowlistRecord(mint, wallet);

    if (!record || !record.active) {
      return { wallet, isAllowlisted: false, record, isExpired: false };
    }

    const now = Math.floor(Date.now() / 1000);
    const isExpired =
      !record.expiresAt.isZero() && record.expiresAt.toNumber() < now;

    return {
      wallet,
      isAllowlisted: !isExpired,
      record,
      isExpired,
    };
  }

  /**
   * Batch-check allowlist status for multiple wallets.
   */
  async batchGetAllowlistStatus(
    mint: PublicKey,
    wallets: PublicKey[]
  ): Promise<AllowlistStatus[]> {
    return Promise.all(wallets.map((w) => this.getAllowlistStatus(mint, w)));
  }
}

// ─── Standalone Helpers ───────────────────────────────────────────────────────

/**
 * Derive the SSS-3 config PDA without instantiating a full client.
 */
export function deriveSss3ConfigPDA(
  mint: PublicKey,
  programId: PublicKey = PROGRAM_ID
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("sss3-config"), mint.toBuffer()],
    programId
  );
}

/**
 * Derive the SSS-3 allowlist record PDA for a wallet.
 */
export function deriveSss3AllowlistPDA(
  mint: PublicKey,
  wallet: PublicKey,
  programId: PublicKey = PROGRAM_ID
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("sss3-allowlist"), mint.toBuffer(), wallet.toBuffer()],
    programId
  );
}

/**
 * Generate a commitment hash for confidential mint verification.
 * Pure function — no network calls required.
 *
 * @example
 * const { commitmentHash, nonce } = generateCommitmentHash(recipient, amount);
 * // pass commitmentHash to confidentialMintSss3()
 * // store nonce off-chain for verification
 */
export function generateCommitmentHash(
  recipient: PublicKey,
  amount: BN,
  nonce?: Buffer
): ConfidentialMintProof {
  const resolvedNonce = nonce ?? crypto.randomBytes(32);
  const amountBuf = amount.toArrayLike(Buffer, "le", 8);

  const hash = crypto.createHash("sha256")
    .update(recipient.toBuffer())
    .update(amountBuf)
    .update(resolvedNonce)
    .digest();

  return {
    commitmentHash: new Uint8Array(hash),
    nonce: new Uint8Array(resolvedNonce),
  };
}

/**
 * Verify a commitment hash against its components.
 * Used by auditors to validate on-chain commitment records.
 */
export function verifyCommitmentHash(
  commitmentHash: Uint8Array,
  recipient: PublicKey,
  amount: BN,
  nonce: Uint8Array
): boolean {
  const amountBuf = amount.toArrayLike(Buffer, "le", 8);
  const expected = crypto.createHash("sha256")
    .update(recipient.toBuffer())
    .update(amountBuf)
    .update(Buffer.from(nonce))
    .digest();

  return Buffer.from(commitmentHash).equals(expected);
}

File: programs/solana-stablecoin-standard/src/instructions/sss3.rs
/// SSS-3: Private Stablecoin Preset
///
/// Experimental preset combining:
///   - Scoped allowlists: only allowlisted addresses may receive transfers
///   - Confidential transfer extension: Token-2022 ElGamal-based privacy layer
///
/// Architecture:
///   - Separate allowlist registry from SSS-2 KYC whitelist
///   - Confidential transfer extension enabled at mint initialization
///   - Transfer hook validates recipient is on allowlist before ciphertext decryption
///
/// Status: Proof-of-concept — Token-2022 confidential transfer tooling is maturing.
///   The allowlist gating is fully functional; confidential amounts use
///   Token-2022's `ConfidentialTransfer` extension which requires client-side
///   ElGamal proof generation (see packages/sdk/src/sss3.ts).
use anchor_lang::prelude::*;
use anchor_spl::token_2022::{self, Token2022};
use anchor_spl::associated_token::AssociatedToken;

use crate::errors::SssError;
use crate::events::{StablecoinInitialized, AllowlistUpdated, ConfidentialMintInitiated};
use crate::state::*;

// ─── Initialize SSS-3 ─────────────────────────────────────────────────────────

#[derive(Accounts)]
#[instruction(params: Sss3Params)]
pub struct InitializeSss3<'info> {
    #[account(mut)]
    pub payer: Signer<'info>,

    /// The new token mint (must be created with ConfidentialTransfer extension pre-initialized)
    #[account(mut)]
    pub mint: Signer<'info>,

    /// SSS config PDA (shared structure with SSS-1/SSS-2)
    #[account(
        init,
        payer = payer,
        space = StablecoinConfig::LEN,
        seeds = [b"sss-config", mint.key().as_ref()],
        bump,
    )]
    pub config: Account<'info, StablecoinConfig>,

    /// SSS-3 specific config PDA
    #[account(
        init,
        payer = payer,
        space = Sss3Config::LEN,
        seeds = [b"sss3-config", mint.key().as_ref()],
        bump,
    )]
    pub sss3_config: Account<'info, Sss3Config>,

    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token2022>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub rent: Sysvar<'info, Rent>,
}

pub fn handler_sss3(ctx: Context<InitializeSss3>, params: Sss3Params) -> Result<()> {
    let config = &mut ctx.accounts.config;
    let sss3_config = &mut ctx.accounts.sss3_config;
    let clock = Clock::get()?;

    // Initialize Token-2022 mint with freeze authority
    // NOTE: Confidential transfer extension must be initialized by the client
    // before calling this instruction (using spl-token CLI or SDK helper).
    // See packages/sdk/src/sss3.ts#initializeConfidentialMint()
    let cpi_ctx = CpiContext::new(
        ctx.accounts.token_program.to_account_info(),
        token_2022::InitializeMint2 {
            mint: ctx.accounts.mint.to_account_info(),
        },
    );
    token_2022::initialize_mint2(
        cpi_ctx,
        params.decimals,
        &ctx.accounts.payer.key(),
        Some(&ctx.accounts.payer.key()), // freeze authority for compliance seizure
    )?;

    // Populate base config
    config.preset = Preset::Sss3;
    config.mint = ctx.accounts.mint.key();
    config.circulating_supply = 0;
    config.supply_cap = params.supply_cap;
    config.mint_authority = ctx.accounts.payer.key();
    config.burn_authority = ctx.accounts.payer.key();
    config.admin = ctx.accounts.payer.key();
    config.pending_admin = None;
    config.paused = false;
    config.decimals = params.decimals;
    config.name = params.name;
    config.symbol = params.symbol;
    config.created_at = clock.unix_timestamp;
    config.total_minted = 0;
    config.total_burned = 0;
    config.bump = ctx.bumps.config;

    // Populate SSS-3 specific config
    sss3_config.mint = ctx.accounts.mint.key();
    sss3_config.allowlist_authority = params.allowlist_authority;
    sss3_config.require_allowlist_for_receive = params.require_allowlist_for_receive;
    sss3_config.require_allowlist_for_send = params.require_allowlist_for_send;
    sss3_config.confidential_transfers_enabled = params.confidential_transfers_enabled;
    sss3_config.auto_approve_new_accounts = params.auto_approve_new_accounts;
    sss3_config.auditor_pubkey = params.auditor_pubkey;
    sss3_config.allowlist_count = 0;
    sss3_config.bump = ctx.bumps.sss3_config;

    emit!(StablecoinInitialized {
        mint: config.mint,
        preset: 3,
        admin: config.admin,
        supply_cap: config.supply_cap,
        timestamp: clock.unix_timestamp,
    });

    Ok(())
}

// ─── Add to Allowlist ─────────────────────────────────────────────────────────

#[derive(Accounts)]
pub struct AllowlistAdd<'info> {
    #[account(mut)]
    pub allowlist_authority: Signer<'info>,

    #[account(
        seeds = [b"sss-config", sss3_config.mint.as_ref()],
        bump = config.bump,
        constraint = config.is_sss3() @ SssError::NotSss3,
        constraint = !config.paused @ SssError::Paused,
    )]
    pub config: Account<'info, StablecoinConfig>,

    #[account(
        seeds = [b"sss3-config", sss3_config.mint.as_ref()],
        bump = sss3_config.bump,
        constraint = sss3_config.allowlist_authority == allowlist_authority.key() @ SssError::Unauthorized,
    )]
    pub sss3_config: Account<'info, Sss3Config>,

    /// The wallet to add to the allowlist
    /// CHECK: Just an address
    pub wallet: AccountInfo<'info>,

    #[account(
        init,
        payer = allowlist_authority,
        space = AllowlistRecord::LEN,
        seeds = [b"sss3-allowlist", sss3_config.mint.as_ref(), wallet.key().as_ref()],
        bump,
    )]
    pub allowlist_record: Account<'info, AllowlistRecord>,

    pub system_program: Program<'info, System>,
}

pub fn handler_allowlist_add(
    ctx: Context<AllowlistAdd>,
    expiry: i64,
    note: [u8; 64],
) -> Result<()> {
    let record = &mut ctx.accounts.allowlist_record;
    let clock = Clock::get()?;

    record.mint = ctx.accounts.sss3_config.mint;
    record.wallet = ctx.accounts.wallet.key();
    record.added_at = clock.unix_timestamp;
    record.expires_at = expiry;
    record.added_by = ctx.accounts.allowlist_authority.key();
    record.active = true;
    record.note = note;
    record.bump = ctx.bumps.allowlist_record;

    // Increment allowlist counter
    ctx.accounts.sss3_config.to_account_info(); // ensure it's loaded
    let sss3_config = &mut ctx.accounts.sss3_config;
    sss3_config.allowlist_count = sss3_config
        .allowlist_count
        .checked_add(1)
        .ok_or(SssError::Overflow)?;

    emit!(AllowlistUpdated {
        mint: record.mint,
        wallet: record.wallet,
        action: 1, // added
        actor: ctx.accounts.allowlist_authority.key(),
        timestamp: clock.unix_timestamp,
    });

    Ok(())
}

// ─── Remove from Allowlist ───────────────────────────────────────────────────

#[derive(Accounts)]
pub struct AllowlistRemove<'info> {
    pub allowlist_authority: Signer<'info>,

    #[account(
        seeds = [b"sss-config", sss3_config.mint.as_ref()],
        bump = config.bump,
        constraint = config.is_sss3() @ SssError::NotSss3,
    )]
    pub config: Account<'info, StablecoinConfig>,

    #[account(
        seeds = [b"sss3-config", sss3_config.mint.as_ref()],
        bump = sss3_config.bump,
        constraint = sss3_config.allowlist_authority == allowlist_authority.key() @ SssError::Unauthorized,
    )]
    pub sss3_config: Account<'info, Sss3Config>,

    /// CHECK: Just an address
    pub wallet: AccountInfo<'info>,

    #[account(
        mut,
        seeds = [b"sss3-allowlist", sss3_config.mint.as_ref(), wallet.key().as_ref()],
        bump = allowlist_record.bump,
        constraint = allowlist_record.active @ SssError::NotAllowlisted,
    )]
    pub allowlist_record: Account<'info, AllowlistRecord>,
}

pub fn handler_allowlist_remove(ctx: Context<AllowlistRemove>) -> Result<()> {
    ctx.accounts.allowlist_record.active = false;

    let clock = Clock::get()?;
    emit!(AllowlistUpdated {
        mint: ctx.accounts.sss3_config.mint,
        wallet: ctx.accounts.wallet.key(),
        action: 0, // removed
        actor: ctx.accounts.allowlist_authority.key(),
        timestamp: clock.unix_timestamp,
    });

    Ok(())
}

// ─── SSS-3 Transfer (Allowlist-gated) ────────────────────────────────────────

/// Standard (non-confidential) transfer with allowlist enforcement.
/// For confidential amounts, use the client-side confidential_transfer instruction
/// in packages/sdk/src/sss3.ts which generates the required ZK proofs.
#[derive(Accounts)]
pub struct TransferSss3<'info> {
    #[account(mut)]
    pub sender: Signer<'info>,

    #[account(
        mut,
        seeds = [b"sss-config", config.mint.as_ref()],
        bump = config.bump,
        constraint = config.is_sss3() @ SssError::NotSss3,
        constraint = !config.paused @ SssError::Paused,
    )]
    pub config: Account<'info, StablecoinConfig>,

    #[account(
        seeds = [b"sss3-config", config.mint.as_ref()],
        bump = sss3_config.bump,
    )]
    pub sss3_config: Account<'info, Sss3Config>,

    /// CHECK: Validated via config
    pub mint: AccountInfo<'info>,

    #[account(mut)]
    /// CHECK: Validated by token program
    pub sender_token_account: AccountInfo<'info>,

    #[account(mut)]
    /// CHECK: Validated by token program
    pub recipient_token_account: AccountInfo<'info>,

    /// Sender's allowlist record (required if require_allowlist_for_send)
    pub sender_allowlist: Option<Account<'info, AllowlistRecord>>,
    /// Recipient's allowlist record (required if require_allowlist_for_receive)
    pub recipient_allowlist: Option<Account<'info, AllowlistRecord>>,

    pub token_program: Program<'info, Token2022>,
}

pub fn handler_transfer_sss3(ctx: Context<TransferSss3>, amount: u64) -> Result<()> {
    require!(amount > 0, SssError::InvalidAmount);

    let sss3 = &ctx.accounts.sss3_config;
    let clock = Clock::get()?;

    // Enforce allowlist for sender
    if sss3.require_allowlist_for_send {
        let record = ctx
            .accounts
            .sender_allowlist
            .as_ref()
            .ok_or(SssError::SenderNotAllowlisted)?;
        require!(record.active, SssError::SenderNotAllowlisted);
        if record.expires_at > 0 {
            require!(record.expires_at > clock.unix_timestamp, SssError::AllowlistExpired);
        }
    }

    // Enforce allowlist for recipient
    if sss3.require_allowlist_for_receive {
        let record = ctx
            .accounts
            .recipient_allowlist
            .as_ref()
            .ok_or(SssError::RecipientNotAllowlisted)?;
        require!(record.active, SssError::RecipientNotAllowlisted);
        if record.expires_at > 0 {
            require!(record.expires_at > clock.unix_timestamp, SssError::AllowlistExpired);
        }
    }

    let cpi_ctx = CpiContext::new(
        ctx.accounts.token_program.to_account_info(),
        token_2022::Transfer {
            from: ctx.accounts.sender_token_account.to_account_info(),
            to: ctx.accounts.recipient_token_account.to_account_info(),
            authority: ctx.accounts.sender.to_account_info(),
        },
    );
    token_2022::transfer(cpi_ctx, amount)?;

    Ok(())
}

// ─── Confidential Mint Initiation ─────────────────────────────────────────────

/// Records the initiation of a confidential mint on-chain.
/// Actual confidential mint requires client-side ElGamal proof generation.
/// This instruction validates authority and emits an auditable event.
#[derive(Accounts)]
pub struct ConfidentialMintSss3<'info> {
    pub mint_authority: Signer<'info>,

    #[account(
        mut,
        seeds = [b"sss-config", config.mint.as_ref()],
        bump = config.bump,
        constraint = config.is_sss3() @ SssError::NotSss3,
        constraint = !config.paused @ SssError::Paused,
        constraint = config.mint_authority == mint_authority.key() @ SssError::Unauthorized,
    )]
    pub config: Account<'info, StablecoinConfig>,

    #[account(
        seeds = [b"sss3-config", config.mint.as_ref()],
        bump = sss3_config.bump,
    )]
    pub sss3_config: Account<'info, Sss3Config>,

    /// Recipient's allowlist record — must be present and active
    pub recipient_allowlist: Account<'info, AllowlistRecord>,
}

pub fn handler_confidential_mint(
    ctx: Context<ConfidentialMintSss3>,
    commitment_hash: [u8; 32],
) -> Result<()> {
    let record = &ctx.accounts.recipient_allowlist;
    let clock = Clock::get()?;

    require!(record.active, SssError::RecipientNotAllowlisted);
    if record.expires_at > 0 {
        require!(record.expires_at > clock.unix_timestamp, SssError::AllowlistExpired);
    }

    // Track supply (amount is concealed — tracked via commitment)
    // In production: verifier program decrypts via auditor ElGamal key
    emit!(ConfidentialMintInitiated {
        mint: ctx.accounts.config.mint,
        recipient_allowlist_record: ctx.accounts.recipient_allowlist.key(),
        commitment_hash,
        initiator: ctx.accounts.mint_authority.key(),
        timestamp: clock.unix_timestamp,
    });

    Ok(())
}
